gradio/upload.py

```diff
@@ -0,0 +1,173 @@
+        return
+    notebook_file = None
+    if app_file.endswith(".ipynb"):
+        notebook_file = app_file
+        app_file = "_" + notebook_file.replace(".ipynb", ".py")
+
+    EMOJI_SET = "🤯🤖🧠🐶👑💥🐮🎁🐙🦋"
+    default_emoji = random.choice(EMOJI_SET)
+    emoji = input(f"Enter Spaces Card emoji [{default_emoji}]: ") or default_emoji
+
+    COLOR_SET = ["red", "yellow", "green", "blue", "indigo", "purple", "pink", "gray"]
+    default_color = random.choice(COLOR_SET)
+    color = input(f"Enter Spaces Card color [{default_color}]: ") or default_color
+
+    readme_header = f"""---
+title: {app_name}
+app_file: {app_file}
+emoji: {emoji}
+colorFrom: {color}
+colorTo: {color}
+sdk: gradio
+{f'notebook_file: {notebook_file}' if notebook_file else ''}
+---
+"""
+    with open(readme_file, "r") as f:
+        readme = f.read()
+    with open(readme_file, "w") as f:
+        f.write(readme_header + readme)
+
+
+def notebook_to_script(notebook_path, script_path):
+    print(f"Converting {notebook_path} to {script_path}.")
+    dependencies = []
+    with open(notebook_path, "r", encoding="utf-8") as f:
+        notebook = json.load(f)
+    import_subprocess = False
+    script_content = []
+
+    for cell in notebook["cells"]:
+        if cell["cell_type"] == "code":
+            for line in cell["source"]:
+                spaces = ""
+                if line.startswith(" ") or line.startswith("\t"):
+                    spaces = line[: len(line) - len(line.lstrip())]
+                    line = line.strip()
+                if line.startswith("!pip install") or line.startswith("%pip install"):
+                    dependencies.extend(line[13:].strip().split(" "))
+                elif line.startswith("%"):
+                    continue
+                elif line.startswith("!"):
+                    command = line[1:].rstrip()
+                    subprocess_call = f"subprocess.run('{command}', shell=True, check=True, text=True)"
+                    script_content.append(spaces + subprocess_call)
+                    import_subprocess = True
+                else:
+                    script_content.append(spaces + line)
+
+            script_content.append("\n")
+
+    if import_subprocess:
+        script_content.insert(0, "import subprocess\n")
+
+    script = "".join(script_content)
+    with open(script_path, "w", encoding="utf-8") as f:
+        f.write(script)
+    return dependencies
+
+
+def upload_to_spaces():
+    if os.getenv("SYSTEM") == "spaces":  # in case a notebook has an upload call
+        return
+
+    hf_api = huggingface_hub.HfApi()
+    whoami = None
+    login = False
+    try:
+        whoami = hf_api.whoami()
+        if whoami["auth"]["accessToken"]["role"] != "write":
+            login = True
+    except:
+        login = True
+    if login:
+        print("Need 'write' access token to create a Spaces repo.")
+        huggingface_hub.login(add_to_git_credential=False)
+        whoami = hf_api.whoami()
```

Comment:
Thanks for the great review @Wauplin and agree with your suggestion: **it's important IMO to reuse as much as possible the logic from `huggingface_hub`** otherwise we would be reinventing the wheel and increase brittleness

Reply:
> it's important IMO to reuse as much as possible the logic from huggingface_hub otherwise we would be reinventing the wheel and increase brittleness

Just to be clear, there is no existing logic for this in hf_hub library, correct? and that @Wauplin was suggesting an api within hf_hub that does not yet exist?

Reply:
Yes exact. Would you mind opening a PR to integrate your logic in `huggingface_hub`? :pray: 

For what I see, `overwrite: bool=True` and  `write_permission: bool=False` should be added to `login`, `notebook_login`, `interpreter_login` and `_login` [in this file](https://github.com/huggingface/huggingface_hub/blob/main/src/huggingface_hub/_login.py#L39).

Reply:
@Wauplin https://github.com/huggingface/huggingface_hub/pull/1476/